design/baremetal-operator/hostdata_crd.md

```diff
@@ -0,0 +1,232 @@
+respectively, and this relationship is controlled via ownerReferences chain. As such,
+the same name and namespace will be used for HostData as its owning object BareMetalHost.
+The spec of HostData will contain the same fields & hardware information as currently
+under `status.hardware` of BareMetalHost.
+
+OwnerReferences refers to the linked BareMetalHost and is added by the operator.
+To see all the available fields of ObjectReference type, check [core/v1](https://pkg.go.dev/k8s.io/api/core/v1#ObjectReference) package.
+
+There will be duplication of inspection data, i.e.,in the HostData spec and BareMetalHost
+status, for the purpose of avoiding breaking API changes. However, we plan to drop
+[hardware](https://github.com/metal3-io/baremetal-operator/blob/05d12b6768a9989a9a4e61dad6cd1f9e84a6e078/apis/metal3.io/v1alpha1/baremetalhost_types.go#L721)
+field from the status of BareMetalHost in the first version bump of the API.
+
+### Pivoting
+
+To pivot objects excluded from Cluster API chain to a target cluster, one can set
+`clusterctl.cluster.x-k8s.io=""` label on HostData and BareMetalHost CRDs.
+Current status annotation based pivoting workflow will stay as it is. However,
+CAPM3 machine controller will no longer copy `status.hardware` of BareMetalHost
+to the annotation. Because that part will be avialable in the form of HostData CR
+in the target cluster.
+
+### Implementation Details/Notes/Constraints
+
+The life of HostData is dependent on BareMetalHost. When BareMetalHost is entering
+inspection state, the HostData will be created. Failure in reaching inspection state
+or having `inspect.metal3.io: disabled` (without `inspect.metal3.io/hardwaredetails`
+annotation) annotation on BareMetalHost will force the operator to not create the
+HostData. Deleting a BareMetalHost will result in deletion of a HostData respectively.
+
+HostData should be immutable and this will be controlled via [Validating Webhook](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#validatingadmissionwebhook) and with RBAC permissions.
+In case, when [re-inspection](https://github.com/metal3-io/baremetal-operator/blob/master/docs/inspectAnnotation.md) of a `available` BareMetalHost is requested,
+new HostData will be re-created, because updating the old HostData will be
+blocked by the Validating Webhook.
+
+Deprovisioning of the BareMetalHost must not delete the HostData. It can
+be deleted either when BareMetalHost is being deleted or when re-inspection
+is requested. To ensure HostData availability while BareMetalHost is provisioned,
+a finalizer should be set on the HostData to prevent its accidental deletion.
+
+Only operator will be granted with create permissions for HostData and everyone
+else will be given no write permissions. If the BareMetalHost was inspected but
+HostData is missing, operator acts differently depending on the current
+state of the BareMetalHost.
+
+- host is provisioned: enter error state
```

Comment:
Well not having HostData will break pivoting at least, in a sense that all the hosts will be re-provisioned. So target cluster will be off for some time.

Reply:
@fmuyassarov I think you mentioned a fair point. Once published, CRs data can be consumed also by other operators. The decomposition is bringing up a scenario where part of the data could be lost (something not possible before), so the owner should take any action to reconstruct them when they are missing. And if it's not possible to immediately rebuild them (as in the case under discussion, an host already provisioned), the owner should notify somehow such condition.

Reply:
Hi @andfasano . I will just write here also about the discussion we had on the community meeting with @zaneb today.

Since normally operator is going to add a finalizer on the HostData, it won't be deleted by an accident until BareMetalHost still exists. It can happen only internationally by the user, who deletes the finalizer manually to able to delete the HostData. And if that's the case, then it is user's fault as Zane pointed out. So, if user deleted HostData when it should not be, then it is responsibility of the user to recover the cluster before proceeding for example with the pivoting. Otherwise, I think we don't have any corner case that would result in the HostData lost while BMH is provisioned. Perhaps we can add an event for such a scenario just for the sake of the user notification.

As such, I've updated this part of the document. Please let me know if you have any case in your mind. Thanks.

Reply:
The use case I had in mind was when the resource could be deleted unintentionally or due a bug, maybe by another operator (and not necessarily by a human being). Even though the finalizer can greatly reduce such risk, that's a chance theoritically possible thanks to to the decomposition (while before, with a single CR, it was a  "all or nothing" scenario), and ignoring completely such event doesn't appear to me a good choice.
From the previous discussions I realized that in case of an already provisioned host it may be not desired to re-trigger an unwanted inspection, so it seems fully reasonable for me the idea of at least raising an event to properly inform the users (or any other potential automated listener) in such case.